src/DataStreams/finalizeBlock.cpp

```diff
@@ -17,7 +17,10 @@ namespace DB
             {
                 current.type = unfinalized_type->getReturnType();
                 if (current.column)
-                    current.column = typeid_cast<const ColumnAggregateFunction &>(*current.column).convertToValues();
+                {
+                    auto mut_column = (*std::move(current.column)).mutate();
```

Comment:
So we're performing finalization on a transient mutable copy of source column, that we will discard right away? Can we make it more explicit, like, a separate method for finalization of states + the old method that extracts values from states? It would be easier to reason about this code. For now, it's not immediately clear that it is correct -- if the column is not shared, why do any mutations and just not assert ownership and change in place. And if it is shared, we will perform finalization many times instead of one.

Reply:
> So we're performing finalization on a transient mutable copy of source column, that we will discard right away? 

Yes.

> Can we make it more explicit, like, a separate method for finalization of states + the old method that extracts values from states? 

I can add new `finalization` method to `ColumnAggregateFunction`, but don't see the reason. It will make things more complex: `ColumnAggregateFunction` will have two states: `finalized` and `not finalized`, and we will need to check that `convertToValues` was called after finalization.
Also, `finalize` won't be constant too, so we will mutate column anyway.

> if the column is not shared, why do any mutations and just not assert ownership and change in place

If column is not shared, mutation does exactly what you said:
```
    MutablePtr shallowMutate() const
    {
        if (this->use_count() > 1)
            return derived()->clone();
        else
            return assumeMutable();
    }
```

> And if it is shared, we will perform finalization many times instead of one.

Yes, and I think it's ok. Here column is constant, and it is a bad idea change something in constant column. It may be done under `mutex`, but even then we don't know if columns are **logically** same or different. It may be possible that other copy of column is being used to merge aggregate states together. So, such `mutex` must protect any read access to aggregation state data.

But actually, this column almost always is not shared (so, mutation is free). The only counterexample I found is in test below (which also causes race condition on master now).

Reply:
>It will make things more complex: ColumnAggregateFunction will have two states: finalized and not finalized, and we will need to check that convertToValues was called after finalization.

It does already have these two states, only they are not explicit. It would be good if we could merge them together, but I'm not sure how to do it. Why do non-finalized columns leave Aggregator at all? Is it for distributed aggregation, so that we can pass around blocks with non-finalized values?

Reply:
Ah, I know, we serialize non-finalized columns.